lib/firenest/synced_server.ex

```diff
@@ -0,0 +1,471 @@
+  """
+  @callback handle_call(request :: term(), GenServer.from(), state()) ::
+              {:reply, reply, new_state}
+              | {:reply, reply, new_state, timeout() | :hibernate}
+              | {:noreply, new_state}
+              | {:noreply, new_state, timeout() | :hibernate}
+              | {:stop, reason, reply, new_state}
+              | {:stop, reason, new_state}
+            when reply: term(), new_state: state(), reason: term()
+
+  @doc """
+  Invoked to handle simple messages.
+
+  See `c:GenServer.handle_info/2` for explanation of return values.
+  """
+  @callback handle_info(request :: term(), state()) ::
+              {:noreply, new_state}
+              | {:noreply, new_state, timeout() | :hibernate}
+              | {:stop, reason :: term(), new_state}
+            when new_state: state()
+
+  @doc """
+  Invoked to handle messages sent with `remote_send/2` and `remote_broadcast/1`.
+
+  See `c:handle_info/2` for explanation of return values.
+  """
+  @callback handle_remote(request :: term(), from :: Topology.node_ref(), state()) ::
+              {:noreply, new_state}
+              | {:noreply, new_state, timeout() | :hibernate}
+              | {:stop, reason :: term(), new_state}
+            when new_state: state()
+
+  @doc """
+  Invoked when a status of a remote synced server changes.
+
+  When the `:up` signal is delivered, a two-step handshake
+  was exchanged between replicas and the remote replica is
+  ready to receive messages.
+
+  See `c:handle_info/2` for explanation of return values.
+  """
+  @callback handle_replica(:up | :down, Topology.node_ref(), state()) ::
+              {:noreply, new_state}
+              | {:noreply, new_state, timeout() | :hibernate}
+              | {:stop, reason :: term(), new_state}
+            when new_state: state()
+
+  @doc """
+  Invoked when the server is about to exit. It should do any cleanup required.
+
+  See `c:GenServer.terminate/2` for detailed explanation.
+  """
+  @callback terminate(reason, state()) :: term()
+            when reason: :normal | :shutdown | {:shutdown, term()}
+
+  @doc """
+  Invoked in some cases to retrieve a formatted version of the
+  `SyncedServer` status.
+
+  See `c:GenServer.format_status/2` for detailed explanation.
+  """
+  @callback format_status(reason, pdict_and_state :: list()) :: term()
+            when reason: :normal | :terminate
+
+  @doc """
+  Invoked to change the state of the `SyncedServer` when a different
+  version of a module is loaded (hot code swapping) and the state’s
+  term structure should be changed.
+
+  See `c:GenServer.code_change/3` for detailed explanation.
+  """
+  @callback code_change(old_vsn, state :: term(), extra :: term()) ::
+              {:ok, new_state :: term()} | {:error, reason :: term()} | {:down, term()}
+            when old_vsn: term()
+
+  @optional_callbacks [handle_replica: 3, terminate: 2, code_change: 3, format_status: 2]
```

Comment:
Elixir v1.7 will mark all callbacks as optional, except `init`. So I would recommend us to use the same technique here and have `__using__` only define the `child_spec`.